app/uk/gov/hmrc/residencenilratebandcalculator/connectors/SessionConnector.scala

```diff
@@ -18,15 +18,117 @@ package uk.gov.hmrc.residencenilratebandcalculator.connectors
+@Singleton
+class SessionConnector @Inject()(val sessionRepository: SessionRepository) {
+  var funcMap: Map[String, (JsValue, CacheMap) => CacheMap] =
+    Map(
+      Constants.estateHasPropertyId -> ((v, cm) => estateHasProperty(v, cm)),
+      Constants.anyExemptionId -> ((v, cm) => anyExemptionClearance(v, cm)),
+      Constants.anyBroughtForwardAllowanceId -> ((v, cm) => anyBroughtForwardAllowance(v, cm)),
+      Constants.anyDownsizingAllowanceId -> ((v, cm) => anyDownsizingAllowance(v, cm)),
+      Constants.anyAssetsPassingToDirectDescendantsId -> ((v, cm) => anyAssetsPassingToDirectDescendants(v, cm)),
+      Constants.anyBroughtForwardAllowanceOnDisposalId -> ((v, cm) => anyBroughtForwardAllowanceOnDisposal(v, cm)))
+
+  private def clearance[A](key: String, value: A, keysToRemove: Set[String], cacheMap: CacheMap)(implicit wrts: Writes[A]): CacheMap = {
+    val mapWithDeletedKeys = cacheMap copy (data = cacheMap.data.filterKeys(s => !keysToRemove.contains(s)))
+    mapWithDeletedKeys copy (data = mapWithDeletedKeys.data + (key -> Json.toJson(value)))
+  }
+
+  private def estateHasProperty[A](value: A, cacheMap: CacheMap)(implicit wrts: Writes[A]): CacheMap =
+    clearance(Constants.estateHasPropertyId, value, Set(Constants.propertyValueId, Constants.percentageCloselyInheritedId, Constants.anyExemptionId, Constants.propertyValueAfterExemptionId), cacheMap)
+
+  private def anyExemptionClearance[A](value: A, cacheMap: CacheMap)(implicit wrts: Writes[A]): CacheMap =
+    clearance(Constants.anyExemptionId, value, Set(Constants.propertyValueAfterExemptionId), cacheMap)
+
+  private def anyBroughtForwardAllowance[A](value: A, cacheMap: CacheMap)(implicit wrts: Writes[A]): CacheMap =
+    clearance(Constants.anyBroughtForwardAllowanceId, value, Set(Constants.broughtForwardAllowanceId), cacheMap)
+
+  private def anyDownsizingAllowance[A](value: A, cacheMap: CacheMap)(implicit wrts: Writes[A]): CacheMap =
+    clearance(Constants.anyDownsizingAllowanceId,
+     value,
+     Set(
+       Constants.dateOfDisposalId,
+       Constants.valueOfDisposedPropertyId,
+       Constants.anyAssetsPassingToDirectDescendantsId,
+       Constants.assetsPassingToDirectDescendantsId,
+       Constants.anyBroughtForwardAllowanceOnDisposalId,
+       Constants.broughtForwardAllowanceOnDisposalId),
+     cacheMap)
+
+  private def anyAssetsPassingToDirectDescendants[A](value: A, cacheMap: CacheMap)(implicit wrts: Writes[A]): CacheMap =
+    clearance(Constants.anyAssetsPassingToDirectDescendantsId, value,
+      Set(
+        Constants.assetsPassingToDirectDescendantsId,
+        Constants.anyBroughtForwardAllowanceOnDisposalId,
+        Constants.broughtForwardAllowanceOnDisposalId),
+      cacheMap)
+
+  private def anyBroughtForwardAllowanceOnDisposal[A](value: A, cacheMap: CacheMap)(implicit wrts: Writes[A]): CacheMap =
+    clearance(Constants.anyBroughtForwardAllowanceOnDisposalId, value, Set(Constants.anyBroughtForwardAllowanceOnDisposalId), cacheMap)
```

Comment:
Should keysToRemove for this be _broughtForwardAllowanceOnDisposalId_, rather than _anyBroughtForwardAllowanceOnDisposalId_?